### PR TITLE
nat: T538: Add static NAT one-to-one

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Ability to set static NAT (one-to-one) in one rule
Use a separate table 'vyos_static_nat' as SRC/DST rules and
STATIC rules can have the same rule number
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T538

## Component(s) name

nat
## How to test
VyOS configuration
```
set nat static rule 10 destination address '10.0.1.1'
set nat static rule 10 inbound-interface 'eth0'
set nat static rule 10 translation address '192.168.1.1'

set nat static rule 20 destination address '203.0.113.0/24'
set nat static rule 20 inbound-interface 'eth0'
set nat static rule 20 translation address '192.0.2.0/24'

```
nftables rules:
```
vyos@r14# sudo nft list table vyos_static_nat
table ip vyos_static_nat {
	chain PREROUTING {
		type nat hook prerouting priority dstnat; policy accept;
		counter packets 3 bytes 252 jump VYOS_PRE_DNAT_HOOK
		iifname "eth0" ip daddr 10.0.1.1 counter packets 0 bytes 0 dnat to 192.168.1.1 comment "STATIC-NAT-10"
		iifname "eth0" counter packets 3 bytes 252 dnat ip prefix to ip daddr map { 203.0.113.0/24 : 192.0.2.0/24 } comment "STATIC-NAT-20"
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		counter packets 3 bytes 252 jump VYOS_PRE_SNAT_HOOK
		oifname "eth0" counter packets 0 bytes 0 snat to 10.0.1.1 comment "STATIC-NAT-10"
		oifname "eth0" counter packets 0 bytes 0 snat ip prefix to ip saddr map { 192.0.2.0/24 : 203.0.113.0/24 } comment "STATIC-NAT-20"
	}

	chain VYOS_PRE_DNAT_HOOK {
		return
	}

	chain VYOS_PRE_SNAT_HOOK {
		return
	}
}
[edit]
vyos@r14# 
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
Something wrong with postrouting rule 10

**Comment by sever-sever** [comment]
> Something wrong with postrouting rule 10

Fixed
```
vyos@r14# sudo nft list table vyos_static_nat
table ip vyos_static_nat {
	chain PREROUTING {
		type nat hook prerouting priority dstnat; policy accept;
		counter packets 0 bytes 0 jump VYOS_PRE_DNAT_HOOK
		iifname "eth0" ip daddr 10.0.1.1 counter packets 0 bytes 0 dnat to 192.168.1.1 comment "STATIC-NAT-10"
		iifname "eth0" counter packets 0 bytes 0 dnat ip prefix to ip daddr map { 203.0.113.0/24 : 192.0.2.0/24 } comment "STATIC-NAT-20"
	}

	chain POSTROUTING {
		type nat hook postrouting priority srcnat; policy accept;
		counter packets 0 bytes 0 jump VYOS_PRE_SNAT_HOOK
		oifname "eth0" ip saddr 192.168.1.1 counter packets 0 bytes 0 snat to 10.0.1.1 comment "STATIC-NAT-10"
		oifname "eth0" counter packets 0 bytes 0 snat ip prefix to ip saddr map { 192.0.2.0/24 : 203.0.113.0/24 } comment "STATIC-NAT-20"
	}

	chain VYOS_PRE_DNAT_HOOK {
		return
	}

	chain VYOS_PRE_SNAT_HOOK {
		return
	}
}
[edit]
vyos@r14# 
```

**Review by c-po — APPROVED** [review]

File: interface-definitions/nat.xml.in [hunk 1]
```diff
@@ -110,6 +121,78 @@
...
         </children>
       </node>
+      <node name="static">
+        <properties>
+          <help>Static NAT (one-to-one)</help>
+        </properties>
+        <children>
+          <tagNode name="rule">
+            <properties>
+              <help></help>
```
Comment (c-po) [inline interface-definitions/nat.xml.in]: No help here?

File: interface-definitions/nat.xml.in [hunk 2]
```diff
@@ -110,6 +121,78 @@
...
+            </properties>
+            <children>
+              #include <include/generic-description.xml.i>
+              <node name="destination">
+                <properties>
+                  <help>NAT destination parameters</help>
+                </properties>
+                <children>
+                  <leafNode name="address">
```
Comment (c-po) [inline interface-definitions/nat.xml.in]: Can we please make a xml.i file for `address`node?

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/nat.xml.in [hunk 3]
```diff
@@ -110,6 +121,78 @@
...
+                      </valueHelp>
+                      <constraint>
+                        <validator name="ipv4-address"/>
+                        <validator name="ipv4-prefix"/>
+                      </constraint>
+                    </properties>
+                  </leafNode>
+                </children>
+              </node>
+              <leafNode name="inbound-interface">
```
Comment (c-po) [inline interface-definitions/nat.xml.in]: Can we please make a xml.i file for `inbound-interface` node?

File: interface-definitions/nat.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/nat.xml.in]: Done

File: interface-definitions/nat.xml.in (hunk 2 above)
Comment (sever-sever) [reply interface-definitions/nat.xml.in]: Done

File: interface-definitions/nat.xml.in (hunk 3 above)
Comment (sever-sever) [reply interface-definitions/nat.xml.in]: Done

File: interface-definitions/nat.xml.in [hunk 4]
```diff
@@ -110,6 +113,54 @@
...
+                  #include <include/ipv4-address-prefix.xml.i>
+                </children>
+              </node>
+              #include <include/inbound-interface.xml.i>
+              <node name="translation">
+                <properties>
+                  <help>Translation address or prefix</help>
+                </properties>
+                <children>
+                  <leafNode name="address">
```
Comment (c-po) [inline interface-definitions/nat.xml.in]: Any reason to not use `interface-definitions/include/ipv4-address-prefix.xml.i`?

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
My only remaining question is if there is a reason to not have this in source/destination NAT. Is it because it would require two rules? Or is it to have a 1:1 NAT (not PAT) for entire subnets?

File: interface-definitions/nat.xml.in (hunk 4 above)
Comment (sever-sever) [reply interface-definitions/nat.xml.in]: Done

**Comment by sever-sever** [comment]
> My only remaining question is if there is a reason to not have this in source/destination NAT. Is it because it would require two rules? Or is it to have a 1:1 NAT (not PAT) for entire subnets?

As I understand, the goal of the task is to have DNAT/SNAT entries but in one rule. So SNAT/DNAT rules do it in the same way.

Maybe @zdc can add something else or I  don't fully understand the task? :)

For example, In the same way, we can add `nat hairpin` which includes 1 rule instead of 3 rules (in the future)

**Review by c-po — APPROVED** [review]

**Pull request merged**